README.md
<div align="center">
  <img alt="Astro Lane" src="/public/android-chrome-192x192.png" width="70" />
</div>

## tshasan.com ##

Implements the fucntions of the astro framwork, tailwindcss and nodejs.
based off the astro lane template and astro paper template




CHANGELOG.md
## 1.0.5

Stable

-   added fixes image error should be fixed
    this should just be version 1 unless i change the api

## 0.0.4.1

-   added a thanksgiving blogpost

## 0.0.4

-added some small fixs to add combatiblity for mobile version

## 0.0.3

Stable
Clean up
-clean up icons logos
-design tweak
-needs better ux or color choice cus its weird
-3d implemented properly

## 0.0.2

alpha version of personal website
-impelement 3d viewer project

## 0.0.1

-blog working


src/config.ts
// Place any global data in this file.
// You can import this data from anywhere in your site by using the `import` keyword.
import type { Site, SocialMediaObjects } from "./types";

export const SITE: Site = {
    siteUrl: "https:/tshasan.com/", // Always put "/" at the end of the URL
    author: "tshasan",
    desc: "This is Taimur Hasan personal portfolio",
    title: "Hello!",
    ogImage: "images/astro-lane.png",
    keywords:
        "Personal portfolio, landing page, page template, developer portfoliom designer portfolio, Taimur, Taimur Hasan, Hasan",
    postPerPage: 3
};

// The site uses iconify - you can find icons on https://iconify.design/.

export const SOCIALS: SocialMediaObjects = [
    {
        name: "Github",
        href: "https://github.com/tshasan",
        icon: "ph:github-logo-duotone",
        title: `Follow ${SITE.author} on Github`,
        active: true
    },
];


astro.config.mjs
import { defineConfig } from 'astro/config';
import { remarkReadingTime } from "./remark-reading-time";
import mdx from "@astrojs/mdx";
import sitemap from "@astrojs/sitemap";
import tailwind from "@astrojs/tailwind";
import react from "@astrojs/react";
import remarkToc from "remark-toc";
import image from "@astrojs/image";
import prefetch from "@astrojs/prefetch";
import { SITE } from "./src/config.ts";
// https://astro.build/config
export default defineConfig({
    site: SITE.siteUrl,
    markdown: {
        syntaxHighlight: "prism",
        remarkPlugins: [remarkToc, remarkReadingTime]
    },
    vite: {
        optimizeDeps: {
            exclude: ["@resvg/resvg-js"]
        },
        ssr: {
            external: ["svgo"]
        }
    },
    integrations: [
        mdx(),
        sitemap(),
        tailwind(),
        image({
            serviceEntryPoint: "@astrojs/image/sharp"
        }),
        prefetch(),
        react()
    ]
});


src/scripts/ascii3d.js
import * as THREE from 'three';
import { AsciiEffect } from 'three/examples/jsm/effects/AsciiEffect.js';
import { TrackballControls } from 'three/examples/jsm/controls/TrackballControls.js';

let camera, controls, scene, renderer, effect;
let mesh; // This will replace 'cube' and 'plane' to be more generic
const start = Date.now()// Define geometries available




const geometries = {
  cube: new THREE.BoxGeometry(200, 200, 200),
  torus: new THREE.TorusGeometry(100, 30, 16, 100),
  prism: new THREE.CylinderGeometry(100, 100, 200, 3), // A simple prism can be represented by a cylinder with 6 sides
  torusknot: new THREE.TorusKnotGeometry( 10, 3, 100, 16 )


};

// Start with the cube geometry
let currentGeometry = geometries.cube;


init();
animate();

function init() {

  const buttonContainer = document.createElement('div');
  buttonContainer.innerHTML = `
    <button id="cubeButton">Cube</button>
    <button id="torusButton">Torus</button>
    <button id="prismButton">Prism</button>
    <button id="torusknotButton">Torus Knot</button>
  `;
  document.body.appendChild(buttonContainer);

  // Add event listeners to buttons
  document.getElementById('cubeButton').addEventListener('click', function() {
    changeGeometry('cube');
  });

  document.getElementById('torusButton').addEventListener('click', function() {
    changeGeometry('torus');
  });

  document.getElementById('prismButton').addEventListener('click', function() {
    changeGeometry('prism');
  });

  document.getElementById('torusknotButton').addEventListener('click', function() {
    changeGeometry('torusknot');
  });

  // Create scence
  
  camera = new THREE.PerspectiveCamera(70, window.innerWidth / window.innerHeight, 1, 1000);
  camera.position.y = 150;
  camera.position.z = 500;

  scene = new THREE.Scene();
  scene.background = new THREE.Color(0, 0, 0);

  const lights = [
    { color: 0xffffff, intensity: 1.5, position: [0, 500, 500] },
    { color: 0xffffff, intensity: 0.5, position: [0, -500, -500] },
  ];

  lights.forEach((light) => {
    const pointLight = new THREE.PointLight(light.color, light.intensity);
    pointLight.position.set(...light.position);
    scene.add(pointLight);
  });

  // Mesh with current geometry
  mesh = new THREE.Mesh(currentGeometry, new THREE.MeshNormalMaterial());
  scene.add(mesh);

  renderer = new THREE.WebGLRenderer();
  renderer.setSize(window.innerWidth, window.innerHeight);

  // Set willReadFrequently to true for performance optimization
  renderer.domElement.style.willReadFrequently = true;

  // ASCII Effect
  effect = new AsciiEffect(renderer, ' .:-+*=%@#', { invert: true });
  effect.setSize(window.innerWidth, window.innerHeight);
  effect.domElement.style.color = 'white';
  effect.domElement.style.backgroundColor = 'black';

  // Append ASCII effect to body
  document.body.appendChild(effect.domElement);

  // Controls
  controls = new TrackballControls(camera, effect.domElement);

  // Handle window resize
  window.addEventListener('resize', onWindowResize, false);

  // Listen for keydown events to change the geometry
  document.addEventListener('keydown', onDocumentKeyDown, false);
}

function onWindowResize() {
  camera.aspect = window.innerWidth / window.innerHeight;
  camera.updateProjectionMatrix();

  renderer.setSize(window.innerWidth, window.innerHeight);
  effect.setSize(window.innerWidth, window.innerHeight);
}

function animate() {
  requestAnimationFrame(animate);
  render();
}

function render() {
  const timer = (Date.now() - start) * 0.0005;

  //mesh.rotation.x = timer;
  //mesh.rotation.y = timer;

  controls.update();
  effect.render(scene, camera);
}

function onDocumentKeyDown(event) {
  var keyCode = event.which;
  // Use keys 1, 2, 3... to switch between geometries
  if (keyCode == 49) { // Key '1'
    changeGeometry('cube');
  } else if (keyCode == 50) { // Key '2'
    changeGeometry('torus');
  } else if (keyCode == 51) { // Key '3'
    changeGeometry('prism');
   } else if (keyCode == 52) { // Key '4'
    changeGeometry('torusknot');
  }
}

function changeGeometry(geometryKey) {
  if (geometries[geometryKey]) {
    // Remove the old mesh from the scene
    scene.remove(mesh);
    // Create a new mesh with the new geometry
    mesh = new THREE.Mesh(geometries[geometryKey], new THREE.MeshNormalMaterial());
    // Add the new mesh to the scene
    scene.add(mesh);
  }
}


src/scripts/asciicube.js
import * as THREE from 'three';
import { AsciiEffect } from 'three/examples/jsm/effects/AsciiEffect.js';
import { TrackballControls } from 'three/examples/jsm/controls/TrackballControls.js';

let camera, controls, scene, renderer, effect;
let cube, plane;
const start = Date.now();

init();
animate();

function init() {
  camera = new THREE.PerspectiveCamera(70, window.innerWidth / window.innerHeight, 1, 1000);
  camera.position.y = 150;
  camera.position.z = 500;

  scene = new THREE.Scene();
  scene.background = new THREE.Color(0, 0, 0);

  const lights = [
    { color: 0xffffff, intensity: 1.5, position: [0, 500, 500] },
    { color: 0xffffff, intensity: 0.5, position: [0, -500, -500] },
  ];

  lights.forEach((light) => {
    const pointLight = new THREE.PointLight(light.color, light.intensity);
    pointLight.position.set(...light.position);
    scene.add(pointLight);
  });

  // Cube
  cube = new THREE.Mesh(
    new THREE.BoxGeometry(200, 200, 200),
    new THREE.MeshNormalMaterial()
  );
  scene.add(cube);

  // Plane
  plane = new THREE.Mesh(
    new THREE.PlaneGeometry(400, 400),
    new THREE.MeshBasicMaterial({ color: 0xe0e0e0 })
  );
  plane.position.y = -200;
  plane.rotation.x = -Math.PI / 2;
  scene.add(plane);

  renderer = new THREE.WebGLRenderer();
  renderer.setSize(window.innerWidth, window.innerHeight);

  // ASCII Effect - Adjust scale for a larger render preview
  effect = new AsciiEffect(renderer, ' .:-+*=%@#', { invert: true, scale: 1 });
  effect.setSize(window.innerWidth, window.innerHeight);
  effect.domElement.style.color = 'white';
  effect.domElement.style.backgroundColor = 'black';

  // Append ASCII effect to body
  document.body.appendChild(effect.domElement);

  // Controls
  controls = new TrackballControls(camera, effect.domElement);

  // Handle window resize
  window.addEventListener('resize', onWindowResize, false);
}

function onWindowResize() {
  camera.aspect = window.innerWidth / window.innerHeight;
  camera.updateProjectionMatrix();

  renderer.setSize(window.innerWidth, window.innerHeight);
  effect.setSize(window.innerWidth, window.innerHeight);
}

function animate() {
  requestAnimationFrame(animate);
  render();
}

function render() {
  const timer = (Date.now() - start) * 0.0005;

  cube.rotation.x = timer;
  cube.rotation.y = timer;

  controls.update();
  effect.render(scene, camera);
}


src/scripts/asciirender.js
import * as THREE from 'three';
import { AsciiEffect } from 'three/examples/jsm/effects/AsciiEffect.js';
import { TrackballControls } from 'three/examples/jsm/controls/TrackballControls.js';

let camera, controls, scene, renderer, effect;
let mesh; // This will replace 'cube' and 'plane' to be more generic
const start = Date.now();

// Define geometries available
const geometries = {
  cube: new THREE.BoxGeometry(200, 200, 200),
  torus: new THREE.TorusGeometry(100, 30, 16, 100),
  prism: new THREE.CylinderGeometry(100, 100, 200, 6), // A simple prism can be represented by a cylinder with 6 sides
  torusknot: new THREE.TorusKnotGeometry( 10, 3, 100, 16 )


};

// Start with the cube geometry
let currentGeometry = geometries.cube;

init();
animate();

function init() {

  const buttonContainer = document.createElement('div');
  buttonContainer.innerHTML = `
    <button id="cubeButton">Cube</button>
    <button id="torusButton">Torus</button>
    <button id="prismButton">Prism</button>
    <button id="torusknotButton">Torus Knot</button>
  `;
  document.body.appendChild(buttonContainer);

  // Add event listeners to buttons
  document.getElementById('cubeButton').addEventListener('click', function() {
    changeGeometry('cube');
  });

  document.getElementById('torusButton').addEventListener('click', function() {
    changeGeometry('torus');
  });

  document.getElementById('prismButton').addEventListener('click', function() {
    changeGeometry('prism');
  });

  document.getElementById('torusknotButton').addEventListener('click', function() {
    changeGeometry('torusknot');
  });
  
  camera = new THREE.PerspectiveCamera(70, window.innerWidth / window.innerHeight, 1, 1000);
  camera.position.y = 150;
  camera.position.z = 500;

  scene = new THREE.Scene();
  scene.background = new THREE.Color(0, 0, 0);

  const lights = [
    { color: 0xffffff, intensity: 1.5, position: [0, 500, 500] },
    { color: 0xffffff, intensity: 0.5, position: [0, -500, -500] },
  ];

  lights.forEach((light) => {
    const pointLight = new THREE.PointLight(light.color, light.intensity);
    pointLight.position.set(...light.position);
    scene.add(pointLight);
  });

  // Mesh with current geometry
  mesh = new THREE.Mesh(currentGeometry, new THREE.MeshNormalMaterial());
  scene.add(mesh);

  renderer = new THREE.WebGLRenderer();
  renderer.setSize(window.innerWidth, window.innerHeight);

  // ASCII Effect
  effect = new AsciiEffect(renderer, ' .:-+*=%@#', { invert: true });
  effect.setSize(window.innerWidth, window.innerHeight);
  effect.domElement.style.color = 'white';
  effect.domElement.style.backgroundColor = 'black';

  // Append ASCII effect to body
  document.body.appendChild(effect.domElement);

  // Controls
  controls = new TrackballControls(camera, effect.domElement);

  // Handle window resize
  window.addEventListener('resize', onWindowResize, false);

  // Listen for keydown events to change the geometry
  document.addEventListener('keydown', onDocumentKeyDown, false);
}

function onWindowResize() {
  camera.aspect = window.innerWidth / window.innerHeight;
  camera.updateProjectionMatrix();

  renderer.setSize(window.innerWidth, window.innerHeight);
  effect.setSize(window.innerWidth, window.innerHeight);
}

function animate() {
  requestAnimationFrame(animate);
  render();
}

function render() {
  const timer = (Date.now() - start) * 0.0005;

  //mesh.rotation.x = timer;
  //mesh.rotation.y = timer;

  controls.update();
  effect.render(scene, camera);
}

function onDocumentKeyDown(event) {
  var keyCode = event.which;
  // Use keys 1, 2, 3... to switch between geometries
  if (keyCode == 49) { // Key '1'
    changeGeometry('cube');
  } else if (keyCode == 50) { // Key '2'
    changeGeometry('torus');
  } else if (keyCode == 51) { // Key '3'
    changeGeometry('prism');
   } else if (keyCode == 52) { // Key '4'
    changeGeometry('torusknot');
  }
}

function changeGeometry(geometryKey) {
  if (geometries[geometryKey]) {
    // Remove the old mesh from the scene
    scene.remove(mesh);
    // Create a new mesh with the new geometry
    mesh = new THREE.Mesh(geometries[geometryKey], new THREE.MeshNormalMaterial());
    // Add the new mesh to the scene
    scene.add(mesh);
  }
}


src/scripts/quoteoftheday.js
document.addEventListener('DOMContentLoaded', function() {
  // Using a CORS Anywhere proxy
  const proxyUrl = 'https://cors-anywhere.herokuapp.com/';
  const targetUrl = 'https://zenquotes.io/api/random';
  
  fetch(proxyUrl + targetUrl)
    .then(response => {
      if (!response.ok) {
        throw new Error('Network response was not ok ' + response.statusText);
      }
      return response.json();
    })
    .then(data => {
      if (data && data.length > 0) {
        const quote = data[0].q;
        const author = data[0].a;
        displayQuote(quote, author);
      }
    })
    .catch(error => {
      console.error('There has been a problem with your fetch operation:', error);
    });
});

function displayQuote(quote, author) {
  const quoteElement = document.getElementById('quoteoftheday');
  if (quoteElement) {
    quoteElement.innerHTML = `"${quote}"<br>— ${author}`;
  } else {
    console.error('No element with ID `quoteoftheday` found.');
  }
}


src/scripts/ThreeRender.js
import * as THREE from "three";
import { AsciiEffect } from "three/examples/jsm/effects/AsciiEffect.js";
import { TrackballControls } from "three/examples/jsm/controls/TrackballControls.js";

// Define variables
let asciiEffectEnabled = true; // Initial state of the ASCII effect
let camera, controls, scene, renderer, effect;
let mesh;

// Define geometries
const geometries = {
    cube: new THREE.BoxGeometry(200, 200, 200),
    torus: new THREE.TorusGeometry(100, 30, 16, 100),
    prism: new THREE.CylinderGeometry(100, 100, 200, 3),
    torusknot: new THREE.TorusKnotGeometry(10, 3, 100, 16)
};

// Set current geometry
let currentGeometry = geometries.cube;

// Initialize and animate
init();
animate();

function init() {
    setupButtons();
    setupScene();
    setupRenderer();
    setupControls();
    window.addEventListener("resize", debounce(onWindowResize, 250), false);
    document.addEventListener("keydown", onDocumentKeyDown, false);
}

function setupScene() {
    camera = new THREE.PerspectiveCamera(
        70,
        window.innerWidth / window.innerHeight,
        1,
        1000
    );
    camera.position.set(0, 150, 500);
    scene = new THREE.Scene();
    scene.background = new THREE.Color(0, 0, 0);
    setupLights();
    const material = new THREE.MeshNormalMaterial({ wireframe: false });
    mesh = new THREE.Mesh(currentGeometry, material);
    scene.add(mesh);
}

function setupLights() {
    const light1 = new THREE.PointLight(0xffffff, 1.5);
    light1.position.set(500, 500, 500);
    scene.add(light1);

    const light2 = new THREE.PointLight(0xffffff, 0.5);
    light2.position.set(-500, -500, -500);
    scene.add(light2);
}

function setupRenderer() {
    renderer = new THREE.WebGLRenderer({ antialias: true }); // Enable antialiasing
    renderer.setSize(window.innerWidth, window.innerHeight);
    effect = new AsciiEffect(renderer, " .:-+*=%@#", { invert: true });
    effect.setSize(window.innerWidth, window.innerHeight);
    const sceneContainer = document.getElementById("scene-container");
    sceneContainer.appendChild(
        asciiEffectEnabled ? effect.domElement : renderer.domElement
    );
}

function setupControls() {
    controls = new TrackballControls(
        camera,
        asciiEffectEnabled ? effect.domElement : renderer.domElement
    );
}

function setupButtons() {
    const sceneContainer = document.getElementById("scene-container");
    const buttonContainer = document.createElement("div");
    buttonContainer.classList.add(
        "w-full", 
        "md:w-1/2",
        "mx-auto",
        "mt-4", 
        "px-4"
    );

    buttonContainer.innerHTML = `
        <button id="cubeButton" class="text-md font-light p-1 text-black bg-gray-200 hover:bg-gray-400 rounded">Cube</button>
        <button id="torusButton" class="text-md font-light p-1 text-black bg-gray-200 hover:bg-gray-400 rounded">Torus</button>
        <button id="prismButton" class="text-md font-light p-1 text-black bg-gray-200 hover:bg-gray-400 rounded">Prism</button>
        <button id="torusknotButton" class="text-md font-light  text-black p-1 bg-gray-200 hover:bg-gray-400 rounded">Torus Knot</button>
        <button id="toggleAsciiButton" class="text-md font-light p-1 text-black bg-gray-200 hover:bg-gray-400 rounded">Toggle ASCII Effect</button>
        <button id="toggleWireframeButton" class="text-md font-light p-1 text-black bg-gray-200 hover:bg-gray-400 rounded">Toggle Wireframe</button>
    `;

    sceneContainer.appendChild(buttonContainer);

    document
        .getElementById("toggleWireframeButton")
        .addEventListener("click", toggleWireframeEffect);
    document
        .getElementById("toggleAsciiButton")
        .addEventListener("click", toggleAsciiEffect);

    buttonContainer.addEventListener("click", (event) => {
        if (event.target.tagName === "BUTTON") {
            const geometryKey = event.target.id.replace("Button", "");
            changeGeometry(geometryKey);
        }
    });
}

function onWindowResize() {
    const container = document.getElementById("scene-container");
    const width = container.clientWidth;
    const height = container.clientHeight;
    camera.aspect = width / height;
    camera.updateProjectionMatrix();
    renderer.setSize(width, height);
    effect.setSize(width, height);
    controls.handleResize();
}

function animate() {
    requestAnimationFrame(animate);
    render();
}

function render() {
    controls.update();
    if (asciiEffectEnabled) {
        effect.render(scene, camera);
    } else {
        renderer.render(scene, camera);
    }
}

function onDocumentKeyDown(event) {
    switch (event.keyCode) {
        case 37: // left arrow key
            mesh.rotation.y -= 0.05;
            break;
        case 38: // up arrow key
            mesh.rotation.x -= 0.05;
            break;
        case 39: // right arrow key
            mesh.rotation.y += 0.05;
            break;
        case 40: // down arrow key
            mesh.rotation.x += 0.05;
            break;
        case 187: // '+' key
        case 107: // numpad '+' key
            camera.fov = Math.max(10, camera.fov - 5);
            camera.updateProjectionMatrix();
            break;
        case 189: // '-' key
        case 109: // numpad '-' key
            camera.fov = Math.min(100, camera.fov + 5);
            camera.updateProjectionMatrix();
            break;
    }
}

function toggleWireframeEffect() {
    mesh.material.wireframe = !mesh.material.wireframe;
}

function toggleAsciiEffect() {
    asciiEffectEnabled = !asciiEffectEnabled;
    const sceneContainer = document.getElementById("scene-container");

    if (asciiEffectEnabled) {
        sceneContainer.replaceChild(effect.domElement, renderer.domElement);
        controls.dispose();
        controls = new TrackballControls(camera, effect.domElement);
    } else {
        sceneContainer.replaceChild(renderer.domElement, effect.domElement);
        controls.dispose();
        controls = new TrackballControls(camera, renderer.domElement);
    }
}

function changeGeometry(geometryKey) {
    if (
        geometries[geometryKey] &&
        currentGeometry !== geometries[geometryKey]
    ) {
        currentGeometry = geometries[geometryKey];
        mesh.geometry.dispose(); // Dispose the old geometry
        mesh.geometry = currentGeometry;
        render(); // Re-render the scene to apply changes
    }
}

function debounce(func, wait, immediate) {
    let timeout;
    return function executedFunction() {
        const context = this;
        const args = arguments;

        const later = function () {
            timeout = null;
            if (!immediate) func.apply(context, args);
        };

        const callNow = immediate && !timeout;
        clearTimeout(timeout);
        timeout = setTimeout(later, wait);

        if (callNow) func.apply(context, args);
    };
}

onWindowResize();
window.addEventListener("resize", onWindowResize, false);
